### PR TITLE
fix: Improve spec compliance by improve handling of responses without content-length or transfer-encoding

### LINKED ISSUE #1414: Undici fails to read bodies from responses with implicit lengths
## Bug Description

Paraphrasing the HTTP spec ([RFC 7230 3.3.3](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.3)), the message length for a normal response body can be defined in a few ways:

* An explicit `content-length` header with a fixed length (point 3 in the RFC list)
* Given a `transfer-encoding` header, a transfer coding chunk that explicitly ends the body (point 5 in the list)
* If otherwise unspecified, all remaining bytes on the connection until the connection is closed (point 7 in the list)

Undici handles the last case incorrectly, making it impossible to read the HTTP response body for this simple "respond and then close the connection" case.

This doesn't come up much for big fancy servers, which tend to use keep-alive etc and explicit framing wherever they can, but it is common on quick simple HTTP server implementations, which avoid state and complexity by just streaming responses and closing the connection when they're done. It's a legitimate way to send responses according to the HTTP spec, and it is supported correctly by browsers and Node's `http` module, but not by Undici.

## Reproducible By

Running in Node 18.0.0:

```javascript
const http = require('node:http');

http.createServer((req, res) => {
    res.removeHeader('transfer-encoding');
    res.writeHead(200, {
        // Header isn't actually necessary, but tells node to close after response
        'connection': 'close'
    });
    res.end('a response body');
}).listen(8008);

fetch('http://localhost:8008').then(async (response) => { // Global fetch from Undici
    console.log('got response', response.status, response.headers);

    try {
        const responseText = await response.text()
        console.log('got body', responseText); // Should log the body OK here
    } catch (e) {
        console.log('error reading body', e); // Throws a 'terminated' error instead
    }
});
```

The server returns a simple response, just containing the default `date` header and a `connection: close` header, then sends the body, then closes the connection.

The `connection: close` header is for clarity - this should work equally well without that, just using `res.socket.end()` explicitly instead.

## Expected Behavior

The above should print the status, headers, and then the response body, which is everything after the headers until the connection is closed.

This does work using fetch in browsers. To test this, run the script above (which will print the fetch error, but then keep running) then load `localhost:8008` in your browser - the string loads successfully.

With that page loaded (for CORS) you can also run the equivalent command in the browser dev console, which also works and prints the response body correctly:

```javascript
fetch('http://localhost:8008').then(async (response) =>
    console.log(await response.text()) // Logs the body correctly, no errors
);
```

This also works using Node's built-in HTTP module:

```javascript
const http = require('http');
const streamConsumers = require('stream/consumers');

http.get('http://localhost:8008', async (res) => {
    const body = await streamConsumers.text(res);
    console.log('GET got body:', body); // Logs the body correctly
});
```

## Logs & Screenshots

Undici's fetch does not print the response body, instead it throws an error:

```javascript
error reading body TypeError: terminated
    at Fetch.onAborted (node:internal/deps/undici/undici:7881:53)
    at Fetch.emit (node:events:527:28)
    at Fetch.terminate (node:internal/deps/undici/undici:7135:14)
    at Object.onError (node:internal/deps/undici/undici:7968:36)
    at Request.onError (node:internal/deps/undici/undici:696:31)
    at errorRequest (node:internal/deps/undici/undici:2774:17)
    at Socket.onSocketClose (node:internal/deps/undici/undici:2236:9)
    at Socket.emit (node:events:527:28)
    at TCP.<anonymous> (node:net:715:12) {
  [cause]: SocketError: closed
      at Socket.onSocketClose (node:internal/deps/undici/undici:2224:35)
      at Socket.emit (node:events:527:28)
      at TCP.<anonymous> (node:net:715:12) {
    code: 'UND_ERR_SOCKET',
    socket: {
      localAddress: undefined,
      localPort: undefined,
      remoteAddress: undefined,
      remotePort: undefined,
      remoteFamily: 'IPvundefined',
      timeout: undefined,
      bytesWritten: 171,
      bytesRead: 90
    }
  }
}
```

## Environment

Ubuntu 
* Node v18.0.0 with built-in fetch
* Node v16.14.2 with Undici 5.1.1

### PR DESCRIPTION
**Pull request by evanderkoogh**
This fixes #1414 #1412 #1490

From [RFC 7230 3.3.3](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.3)
```
       Otherwise, this is a response message without a declared message
       body length, so the message body length is determined by the
       number of octets received prior to the server closing the
       connection.
```

So now if we get a SocketClosed, we check if we have seen a `content-length` or `transfer-encoding` header and if not, we complete the message parsing.

### CODE REVIEW

**Review by ronag — COMMENTED** [review]
Needs a test. Also I need to think about this. Feels like there might be some edge cases here.Especially with pipelining.

**Comment by evanderkoogh** [comment]
It does have a test: https://github.com/nodejs/undici/pull/1540/commits/ac2fbb96e612b6cbb999a956fbe03a66673d937f#diff-ecc2f3b14c967f02fd074ac2d0e0b876e683c84f2755d251bbaa74a813a82130

It is really weird that it says [line 620](https://app.codecov.io/gh/nodejs/undici/compare/1540/diff?src=pr&el=tree#D1L620) is not covered by test because [line 928](https://app.codecov.io/gh/nodejs/undici/compare/1540/diff?src=pr&el=tree#D1L968) can never be executed if `this.hasTransferEncoding` has not been set to true?

**Comment by evanderkoogh** [comment]
Oh and the `Node CI / Test / test (18, ubuntu-latest, experimental) (pull_request) Failing after 2m` failing test is a timeout in some unrelated code.

Ahh.. I think I see where your concern is around the edge cases and pipelining.. I would need to check, but it could be indeed possible that when you are pipelining, it would error out the next request. I am going to investigate.

**Comment by mcollina** [comment]
I think we should just error all the pipelined requests. Given they are GETs, they will be picked up whenever another socket is connected.

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you add a test with more than one `client.request()` executing in parallel with pipelining enabled and disabled? So we verify this works ok in all situations.

**Comment by Guilledll** [comment]
Hi everyone, any news about merging this? It will be very usefull for me to have it, thanks for the great work.

**Comment by mcollina** [comment]
This still needs one more test: https://github.com/nodejs/undici/pull/1540#pullrequestreview-1038547402

@evanderkoogh are you still interesting in completing this PR?

File: lib/client.js [hunk 1]
```diff
@@ -615,6 +616,8 @@ class Parser {
       this.keepAlive += buf.toString()
     } else if (key.length === 14 && key.toString().toLowerCase() === 'content-length') {
       this.contentLength += buf.toString()
+    } else if (key.length === 16 && key.toString().toLowerCase() === 'transfer-encoding') {
```
Comment (kyranet) [inline lib/client.js:619]: Judging by the line 617 (since I'm not very familiar with Undici's code), shouldn't this be `17`? `transfer-encoding` has 3 more characters than the previous header, and `.length` does indeed evaluate to `17`.

Comment (ronag) [reply lib/client.js:619]: Sounds like a test is missing for this?

File: lib/client.js [hunk 2]
```diff
@@ -957,10 +960,16 @@ function onSocketEnd () {
...
-  this[kParser].destroy()
-  this[kParser] = null
+  // without content-length or transfer-encoding header, the body is defined as everything
+  // the server sends before closing the connection. RFC 7230 3.3.3 point 7.
+  if (!parser.contentLength && !parser.hasTransferEncoding) {
+    parser.onMessageComplete()
+  }
+
+  parser.destroy()
+  parser = null
```
Comment (ronag) [inline lib/client.js:972]: This doesn't properly null the parser. Please don't use a local var.

**Review by ronag — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
Would you mind sending a PR finishing this?

**Comment by DamianGlowala** [comment]
Affects me (a Nuxt user) as well; receiving an error when proxying the request through a Nitro server:

image

**Comment by garand** [comment]
😢 So is this not getting fix for now?

**Comment by xgqfrms** [comment]
Any updates?

**Comment by metcoder95** [comment]
PR has been closed for a while, but this seems to be still something to be considered to add. If interested, feel free to open a new PR with a proposal and we can outline something 🙂

**Comment by evanderkoogh** [comment]
The problem is that when I made that change, I fixed the actual issue, but then realised I introduced a bunch of edge-cases where there were still problems and I didn't have any bandwidth to figure out how to solve for all of those.

**Comment by DamianGlowala** [comment]
I would have assumed a non-compliance with [RFC 7230 3.3.3](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.3) should gain a bit more attention from the Node.js team than simply 'feel free to open a new PR'...

Clearly this might be not a trivial problem to solve, which author of this PR already expressed in the above comment, hence any form of *support* from the team would be (very!) appreciated.

**Comment by metcoder95** [comment]
The team has limited bandwidth for addressing all the issues opened at the time, hence we asked the community if they are willing to contribute to fixing the problem meanwhile we provide all the support and guidance through reviews.

Since the PR has been closed for a while, it might be better to open a new one where we can continue the efforts if someone volunteers 🙂
